discovery_engine_core/ai/xayn-ai/src/coi/key_phrase.rs

```diff
@@ -154,50 +172,64 @@ impl RelevanceMap {
 
 /// Unifies the key phrases and candidates.
 fn unify(
-    mut key_phrases: BTreeSet<KeyPhrase>,
+    key_phrases: Vec<Arc<KeyPhrase>>,
     candidates: &[String],
     smbert: impl Fn(&str) -> Result<Embedding, Error> + Sync,
-) -> BTreeSet<KeyPhrase> {
+) -> Vec<Arc<KeyPhrase>> {
     #[cfg(not(feature = "multithreaded"))]
     let candidates = candidates.iter();
     #[cfg(feature = "multithreaded")]
     let candidates = candidates.into_par_iter();
 
-    let mut candidates = candidates
-        .filter(|candidate| !key_phrases.contains(*candidate))
+    let candidates = candidates
+        .filter(|&candidate| {
+            key_phrases
+                .iter()
+                .all(|key_phrase| key_phrase.words() != candidate)
+        })
         .map(clean_key_phrase)
+        .collect::<HashSet<_>>()
+        .into_iter()
         .filter_map(|candidate| {
             smbert(&candidate)
-                .and_then(|point| KeyPhrase::new(candidate, point).map_err(|e| e.into()))
+                .and_then(|point| {
+                    KeyPhrase::new(candidate, point)
+                        .map(Arc::new)
+                        .map_err(|e| e.into())
+                })
                 .ok()
-        })
-        .collect::<BTreeSet<_>>();
+        });
 
-    key_phrases.append(&mut candidates);
-    key_phrases
+    key_phrases.into_iter().chain(candidates).collect()
 }
 
 /// Reduces the matrix along the axis while skipping the diagonal elements.
-fn reduce_without_diag<S, F, G>(
+///
+/// The elements can be prepared before they are reduced. The reduced result can be finalized
+/// whereas the finalization conditially depends on whether the reduced lane is part of the main
+/// square block of the matrix.
+fn reduce_without_diag<S, P, R, F>(
     a: ArrayBase<S, Ix2>,
     axis: Axis,
-    reduce: F,
-    finalize: G,
+    prepare: P,
+    reduce: R,
+    finalize: F,
 ) -> Array2<f32>
 where
     S: Data<Elem = f32>,
-    F: Fn(f32, f32) -> f32 + Copy,
-    G: Fn(f32) -> f32 + Copy,
+    P: Fn(f32) -> f32,
+    R: Fn(f32, f32) -> f32,
+    F: Fn(f32, bool) -> f32,
 {
     a.lanes(axis)
         .into_iter()
         .enumerate()
         .map(|(i, lane)| {
             lane.iter()
                 .enumerate()
-                .filter_map(|(j, element)| (i != j).then(|| *element))
-                .reduce(reduce)
-                .map(finalize)
+                .filter_map(|(j, element)| (i != j).then(|| prepare(*element)))
+                .reduce(|x, y| reduce(x, y))
```

Comment:
this doesn't work because of
```
cannot move out of `reduce`, a captured variable in an `FnMut` closure
move occurs because `reduce` has type `R`, which does not implement the `Copy` trait
```
i removed the `Copy` bounds on the arguments to allow for more general closures than just function pointers. the small downside is that we have to type this slightly more redundant.

Reply:
ok, thanks for checking!